### DISCUSSION TITLE
generazioni certificati https con caddy e duckdns

### DISCUSSION THREAD

**Discussion by enry86cami** [indomus/forum#2503] [HUB: Home Assistant]
Ciao a tutti,
sto migrando all'installazione di HA su docker e sono arrivato al punto della generazione dei certificati, ho seguito la guida dell'installazione di caddy su docker ed è correttamente installato.
uso il servizio DUCKDNS, ma ho il problema che non posso fare il forward delle porte 80 e 443.
Da come ho capito dovrei applicare una DNS CHALLENGE, ma non ho capito come va configurato il caddyfile.
Potrei ricevere aiuto?

**Comment by snakuzzo** [comment]
Ciao.
Non è solo questione di come configurare il Caddyfile. Ti occorre buildare un'immagine con il modulo ad hoc per fare la challenge. Non esiste l'immagine ufficiale già pronta, ma su dockerhub forse la trovi già pronta...preparata da altri.
Se può esserti utile, io ne ho preparata una su ghcr.io con i moduli per duckdns e cloudflare [QUI](https://github.com/snakuzzo/caddy-cloudflare-duckdns)

Una volta buildato il container puoi pensare al Caddyfile.
Questo di seguito è un esempio del Caddyfile utilizzato dal container creato con l'immagine linkata sopra:

```
(https_header) {
	header {
		Strict-Transport-Security "max-age=31536000; includeSubdomains"
		X-XSS-Protection "1; mode=block"
		X-Content-Type-Options "nosniff"
		X-Frame-Options "SAMEORIGIN"
		Referrer-Policy "same-origin"
	}
}

https://tuodominio.duckdns.org:9000 {
	import https_header
	reverse_proxy http://192.168.1.10:8123
        # DuckDNS DNS Challenge
	tls example@gmail.com {
		dns duckdns xxxxxxxxxxxxxxxxxxxxxxxx  # duckdns token
	}
}
```

**Reply by enry86cami** [reply]
grazie per la risposta, purtroppo non sono così abile su questo tema.
Sembra che sono riuscito a sbloccare il forward delle porte 80 e 443, quindi ipotizzo che potrei proseguire passo-passo la guida:
https://indomus.it/guide/come-installare-e-configurare-caddy-con-docker-su-raspbian-di-raspberry-pi/
semplicemente sbloccando la 443, domanda forse stupida, da questo momento la porta 443 sarà occupata sempre da caddy per il rinnovo dei certificati?

```

**Reply by snakuzzo** [reply]
Sì, ma puoi usarla anche per esporre il tuo servizio (sempre tramite caddy). Non è che la usi solo per la challenge tls-alpn01

**Reply by enry86cami** [reply]
ho seguito la guida e ora riesco a collegarmi mediante https://mioduckdns.org:9000.
un dubbio, ora se faccio http://mioduckdns.org:8764, non raggiungo motioneye.... ora passa tutto su caddy e quindi devo inserire su caddyfile anche motioneye?

**Reply by snakuzzo** [reply]
Ma cosa vuol dire?
Non è che se metti Caddy viene tutto proxato da Caddy... dipende da come hai configurato tu le cose. Non puoi fare una domanda del genere se non dettagli cosa hai fatto

**Reply by enry86cami** [reply]
Perdona, avevo dato per scontato avendo fatto copia incolla con guida...

`(https_header) {
  header {
    Strict-Transport-Security "max-age=31536000; includeSubdomains"
    X-XSS-Protection "1; mode=block"
    X-Content-Type-Options "nosniff"
    X-Frame-Options "SAMEORIGIN"
    Referrer-Policy "same-origin"
  }
}

https://casamia.duckdns.org { 
  respond "Saluti dal Crew inDomus"
  file_server
}
https://casamia.duckdns.org:9000 { 
  import https_header
  reverse_proxy http://IP_DELLA_ISTANZA_HOME_ASSISTANT:8123
}`

**Reply by snakuzzo** [reply]
Scusa, ma perché dovrebbe proxarti motioneye? Io vedo solo l'esempio e HA.

**Reply by enry86cami** [reply]
Da quando ho questa configurazione su caddy, http://myip.duckdns.org:8764 non risolve più la pagina ... ERR_TIMED_OUT

**Reply by snakuzzo** [reply]
Caddy non c'entra niente

**Reply by enry86cami** [reply]
capito, però se dall'esterno cerco di raggiungere http://ippubblico:8764 riesco a raggiungere motioneye, invece da http://myip.duckdns.org:8764/ no.

Per questo pensavo che caddy facesse qualcosa....

**Reply by snakuzzo** [reply]
sì ma il discorso è che devi capire cosa hai fatto con Caddy...segui guide, consigli...ma non devi dare nulla per scontato.
Rileggi il Caddyfile...secondo te con quella configurazione cosa stai facendo ?
Stai semplicemente proxando il servizio di HA (che gira sulla porta 8123) in https sulla porta 9000.

Quindi protocollo http e porta 8764 non c'entrano nulla.
Poi, se dici che tramite IP pubblico va e con FQDN  va in timeout (ERR_TIMED_OUT), vuol dire che probabilmente il tuo FQDN non viene risolto con l'IP corretto...fai questa verifica

**Reply by enry86cami** [reply]
sicuramente mi perdo qualcosa ma se da fuori lan faccio "ping myip.duckdns.org", mi risponde l'ip "mypubIP".
da web faccio:
http://mypubIP:8764 e funziona
http://myip.duckdns.org:8764 non va....

sinceramente non saprei...

**Reply by snakuzzo** [reply]
premesso che il ping non è una prova per vedere come viene risolto un domnio dai DNS, perché potrebbe dipendere anche da com'è configurato il file hosts o altre cose.
Questi test si fanno utilizzando `nslookup` o `dig`.

Sicuro di fare la prova dall'esterno e non da dentro la tua stessa LAN ?

**Reply by enry86cami** [reply]
con nslookup ottengo lo stesso IP, sono sicuro in quanto sono fuori casa al momento...
